### ISSUE TITLE
subsampling rate and training iterations

### ISSUE THREAD

**Issue by yhx-03** [DingfanChen/GS-WGAN#12]
Hi @DingfanChen:

Hope you are doing well. I wonder why calculate this privacy budget this way or not like dp-sgd.
The variables that differ from dp-sgd are  prob= 1. / config['num_discriminators'] and coeff=n_steps * batch_size. ^_^
```python
delta = 1e-5
batch_size = config['batchsize']
prob = 1. / config['num_discriminators']  # subsampling rate
n_steps = config['iterations']  # training iterations
sigma = config['noise_multiplier']  # noise scale
func = lambda x: rdp_bank.RDP_gaussian({'sigma': sigma}, x)

acct = rdp_acct.anaRDPacct()
acct.compose_subsampled_mechanism(func, prob, coeff=n_steps * batch_size)
epsilon = acct.get_eps(delta)
print("Privacy cost is: epsilon={}, delta={}".format(epsilon, delta))
```

**Comment by DingfanChen** [comment]
Hi, 
Thank you for bringing up this point. Simply put, our approach involves sanitizing the intermediate gradient (i.e., the gradient w.r.t. each synthetic sample) that is transferred between two modules (the Generator and Discriminator). As a result, the sensitivity and Gaussian mechanism for each step may appear slightly different from the standard case where only one model's gradient is directly sanitized using DP-SGD.

Intuitively, in our case, every individual training sample can contribute to *all* of the single intermediate gradients (which we are sanitizing) within the batch, leading to a ```batch_size``` multiplicative factor. In contrast, in the case of normal DP-SGD training, each individual training sample contributes to only one "per-example gradient" within each batch.

Moreover, in normal DP-SGD training, training samples not in the current batch do not affect the resulting sanitized gradient for the current iteration. However, in our case, even training samples not in the current batch may still influence the resulting gradients. This is because the Discriminator is trained non-privately on all training samples, accumulating information about all samples in its parameters, which in turn affects the intermediate gradients that we are sanitizing. This essentially means that subsampling is only bounded to each sub-discriminator in the worst-case scenario.

In summary, we can only conduct a conservative privacy analysis by treating the transfer of *each single gradient w.r.t. to one synthetic sample* as one Gaussian mechanism (leading to a batch_size factor) and considering subsampling only up to the data partition used for training each sub-discriminator. 

I hope this explanation helps clarify the point.

**Comment by yhx-03** [comment]
Thank you so much! I do appreciate your detailed demonstrations, it means a lot to me!
